### PR TITLE
Support proxy protocol v2

### LINKED ISSUE #4904: Teleport doesn't support proxy protocol v2
### Description

**What happened**: Enable [proxy protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) v2 (binary version) on an AWS NLB. Teleport can't parse the binary version (v2), only the human-readable version (v1), so you get a lot of multiplexer errors:

```
Nov 18 20:12:33 ip-172-31-2-140.ec2.internal /usr/bin/teleport[3496]: WARN [MX:PROXY:] "\nERROR REPORT:\nOriginal Error: *trace.BadParameterError failed to detect protocol by prefix: [13 10 13]\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:348 github.com/gravitational/teleport/lib/multiplexer.detectProto\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:263 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:197 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit\nUser Message: failed to detect protocol by prefix: [13 10 13]\n" multiplexer/multiplexer.go:200
Nov 18 20:12:33 ip-172-31-2-140.ec2.internal /usr/bin/teleport[3496]: WARN [MX:PROXY:] "\nERROR REPORT:\nOriginal Error: *trace.BadParameterError failed to detect protocol by prefix: [13 10 13]\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:348 github.com/gravitational/teleport/lib/multiplexer.detectProto\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:263 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:197 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit\nUser Message: failed to detect protocol by prefix: [13 10 13]\n" multiplexer/multiplexer.go:200
Nov 18 20:12:33 ip-172-31-2-140.ec2.internal /usr/bin/teleport[3496]: WARN [MX:PROXY:] "\nERROR REPORT:\nOriginal Error: *trace.BadParameterError failed to detect protocol by prefix: [13 10 13]\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:348 github.com/gravitational/teleport/lib/multiplexer.detectProto\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:263 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:197 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit\nUser Message: failed to detect protocol by prefix: [13 10 13]\n" multiplexer/multiplexer.go:200
Nov 18 20:12:34 ip-172-31-2-140.ec2.internal /usr/bin/teleport[3496]: WARN [MX:PROXY:] "\nERROR REPORT:\nOriginal Error: *trace.BadParameterError failed to detect protocol by prefix: [13 10 13]\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:348 github.com/gravitational/teleport/lib/multiplexer.detectProto\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:263 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:197 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit\nUser Message: failed to detect protocol by prefix: [13 10 13]\n" multiplexer/multiplexer.go:200
Nov 18 20:12:35 ip-172-31-2-140.ec2.internal /usr/bin/teleport[3496]: WARN [MX:PROXY:] "\nERROR REPORT:\nOriginal Error: *trace.BadParameterError failed to detect protocol by prefix: [13 10 13]\nStack Trace:\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:348 github.com/gravitational/teleport/lib/multiplexer.detectProto\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:263 github.com/gravitational/teleport/lib/multiplexer.detect\n\t/go/src/github.com/gravitational/teleport/lib/multiplexer/multiplexer.go:197 github.com/gravitational/teleport/lib/multiplexer.(*Mux).detectAndForward\n\t/opt/go/src/runtime/asm_amd64.s:1374 runtime.goexit\nUser Message: failed to detect protocol by prefix: [13 10 13]\n" multiplexer/multiplexer.go:200
```

**What you expected to happen**: Proxy protocol v2 headers should be parsed correctly.

**How to reproduce it (as minimally and precisely as possible)**: Deploy Teleport in AWS using [HA reference Terraform](https://github.com/gravitational/teleport/tree/master/examples/aws/terraform/ha-autoscale-cluster), enable proxy protocol for proxyweb group, watch errors.

### Environment
- Teleport version (use `teleport version`): `Teleport Enterprise v4.4.5 git:v4.4.5-0-g23a2e42a8 go1.14.4`

gz#4292

### PR DESCRIPTION
**Pull request by probakowski**
This change add multiplexer support for [proxy protocol v2](https://www.haproxy.org/download/2.6/doc/proxy-protocol.txt) (binary). 
I've tested it manually using local HAProxy instance with config:

```
backend teleport
    balance leastconn
    mode tcp
    retries 3
    option redispatch
    server localhost:3080 127.0.0.1:3080 send-proxy-v2

frontend proxy
    bind  :5816
    log global
    mode tcp
    default_backend teleport

    maxconn 60000
    timeout client 10m
```
and then:
```sh
tsh --proxy=localhost:5816 login
```

Closes https://github.com/gravitational/teleport/issues/4904

### CODE REVIEW

File: lib/multiplexer/multiplexer_test.go [hunk 1]
```diff
@@ -189,6 +189,57 @@ func TestMux(t *testing.T) {
 		require.Equal(t, out, remoteAddr.String())
 	})
 
+	// ProxyLine tests proxy line protocol
+	t.Run("ProxyLineV2", func(t *testing.T) {
+		listener, err := net.Listen("tcp", "127.0.0.1:0")
+		require.Nil(t, err)
```
Comment (zmb3) [inline lib/multiplexer/multiplexer_test.go:195]: nit: `require.NoError` reads a little bit nicer.

File: lib/multiplexer/proxyline.go [hunk 2]
```diff
@@ -122,3 +127,72 @@ func parseIP(protocol string, addrString string) (net.IP, error) {
...
+		return nil, trace.BadParameter("unrecognized signature %s", hex.EncodeToString(header.Signature[:]))
+	}
+	cmd, ver := header.VersionCommand&0xF, header.VersionCommand>>4
+	if ver != 2 {
+		return nil, trace.BadParameter("unsupported version %d", ver)
+	}
+	if cmd == 0 {
+		// LOCAL command, just skip address information and keep original addresses (no proxy line)
+		if header.Length > 0 {
+			_, err := io.CopyN(ioutil.Discard, reader, int64(header.Length))
```
Comment (zmb3) [inline lib/multiplexer/proxyline.go:168]:
```suggestion
			_, err := io.CopyN(io.Discard, reader, int64(header.Length))
```

File: lib/multiplexer/proxyline.go [hunk 3]
```diff
@@ -122,3 +127,72 @@ func parseIP(protocol string, addrString string) (net.IP, error) {
...
+	var header proxyV2Header
+	var ret ProxyLine
+	if err := binary.Read(reader, binary.BigEndian, &header); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if bytes.Compare(header.Signature[:], proxyV2Prefix) != 0 {
+		return nil, trace.BadParameter("unrecognized signature %s", hex.EncodeToString(header.Signature[:]))
+	}
+	cmd, ver := header.VersionCommand&0xF, header.VersionCommand>>4
+	if ver != 2 {
```
Comment (zmb3) [inline lib/multiplexer/proxyline.go:162]: Can we add constants for these magic numbers?
- version 2
- cmd 0
- cmd 1
- protocol 0x11
- protocol 0x21

**Review by ibeckermayer — APPROVED** [review]
Don't have the context to review how it works, but the code itself LGTM after addressing Zac's comments.

File: lib/multiplexer/proxyline.go (hunk 3 above)
Comment (probakowski) [reply lib/multiplexer/proxyline.go]: Added constants

File: lib/multiplexer/proxyline.go (hunk 2 above)
Comment (probakowski) [reply lib/multiplexer/proxyline.go]: Done

**Review by zmb3 — APPROVED** [review]

**Comment by probakowski** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|branch/v9|[](https://github.com/gravitational/teleport/pull/11722)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**